llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,443 @@
+
+  bool isNotLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                    const MachineBasicBlock &MBB);
+
+  bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                    SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+  bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                              const MachineInstr *InstrB);
+
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+};
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  bool Modified = renameMBBPhysRegs(LoopBBs);
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(
+    const SmallVector<const MachineBasicBlock *, 4> &LoopBBs) {
+  bool Modified = false;
+  for (const MachineBasicBlock *MBB : LoopBBs) {
+    LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+               dbgs() << "\n");
+    SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+    SmallSet<MCPhysReg, 16> ForbiddenPhysRegs = getBBForbiddenReplacements(MBB);
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isIdentityCopy(MI))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Op.getReg().isVirtual())
+          continue;
+        if (VRM->hasRequiredPhys(Op.getReg()))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI->isVecOrAccRegClass(MRI->getRegClass(Reg)))
+          continue;
+        //  Subregisters can reuse Constants in the high part of the
+        //  register, therefore renaming a reused physical register can
+        //  prohibit the elimination of a move of constants, thus degrading
+        //  performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse))
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs)) {
+          MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+          bool Replaced = replaceReg(Reg, ForbiddenPhysRegs);
+
+          if (Replaced) {
```

Comment:
nit: as we have just one use for `Raplaced`, we can use directly `if (replaceReg(...))`. Or just `const bool Replaced =...`

Reply:
folded it into the if